src/hotspot/share/classfile/dictionary.cpp

```diff
@@ -450,18 +446,17 @@ void Dictionary::clean_cached_protection_domains() {
             ls.print(" loading: "); probe->instance_klass()->print_value_on(&ls);
             ls.cr();
           }
-          if (probe->pd_set() == current) {
-            probe->set_pd_set(current->next());
+          if (probe->pd_set_acquire() == current) {
+            probe->release_set_pd_set(current->next_acquire());
           } else {
             assert(prev != NULL, "should be set by alive entry");
-            prev->set_next(current->next());
+            prev->release_set_next(current->next_acquire());
           }
-          ProtectionDomainEntry* to_delete = current;
-          current = current->next();
-          delete to_delete;
+          delete_list->push(current);
```

Comment:
Suggested comment before this:
// Mark current for deletion, but in the meantime it can still be traversed

It is important that current is not unlinked so that in-progress traversals do not break. Though to be honest I'm unclear exactly what triggers the clearing of an entry this way.

Reply:
I added the comment.  The service thread could be trying to remove this entry and there's a small chance that some class loading will be looking up this entry with SystemDictionary::find_instance_klass.

    // First clean cached pd lists in loaded CLDs
    // It's unlikely, but some loaded classes in a dictionary might
    // point to a protection_domain that has been unloaded.
    // The dictionary pd_set points at entries in the ProtectionDomainCacheTable.

There's a test runtime/Dictionary/ProtectionDomainCache.java that does this. It creates a protection domain with a jar file loading with a URL class loader that's then removed.  The defining class loader of the class loaded has that protection domain in its pd_set.